hands-on-exercise/4-run-in-prod/3-add-leaderboard.md

```diff
@@ -0,0 +1,1640 @@
+
+```diff-go [https://github.com/cosmos/b9-checkers-academy-draft/blob/leaderboard-objects/x/leaderboard/types/params.go#L14]
+    var (
+        KeyLength = []byte("Length")
+-      // TODO: Determine the default value
+-      DefaultLength uint64 = 0
++  )
++
++  const (
++      DefaultLength uint64 = 100
+)
+```
+
+You also make it `const` as this is the case and comes handy later on.
+
+## New v2 information
+
+It is time to take a closer look at the new data structures being introduced with the new module.
+
+<HighlightBox type="tip">
+
+If you feel unsure about creating new data structures with Ignite CLI, look at the [previous sections](/hands-on-exercise/1-ignite-cli/4-create-message.md) of the exercise again.
+
+</HighlightBox>
+
+To give the new v2 information a data structure, you need the following:
+
+1. Add a structure for **the leaderboard**: you want a single stored leaderboard for the whole module. Let Ignite CLI help you implement a structure:
+
+    <CodeGroup>
+
+    <CodeGroupItem title="Local" active>
+
+    ```sh
+    $ ignite scaffold single leaderboard winners \
+        --module leaderboard --no-message
+    ```
+
+    </CodeGroupItem>
+
+    <CodeGroupItem title="Docker">
+
+    ```sh
+    $ docker run --rm -it \
+        -v $(pwd):/checkers \
+        -w /checkers \
+        checkers_i \
+        ignite scaffold single leaderboard winners \
+        --module leaderboard --no-message
+    ```
+
+    </CodeGroupItem>
+
+    </CodeGroup>
+
+2. This creates a Protobuf file with `string winners`. This is not very useful. So you declare by hand another Protobuf message in `leaderboard.proto` for use as a leaderboard rung:
+
+    ```diff-protobuf [https://github.com/cosmos/b9-checkers-academy-draft/blob/leaderboard-object/proto/leaderboard/leaderboard.proto#L11-L15]
+        message Leaderboard {
+            string winners = 1;
+        }
+
+    +  message Winner {
+    +      string address = 1;
+    +      uint64 wonCount = 2;
+    +      uint64 addedAt = 3;
+    +  }
+    ```
+
+    <HighlightBox type="note">
+
+    Where:
+
+    * `address` indicates the player. It will be the same address as the one that comes in `PlayerInfo.index`.
+    * `wonCount` determines the ranking on the leaderboard - the higher the count, the closer to the `0` index in the array. This should exactly match the value found in the corresponding player stats. This duplication of data is a lesser evil because if `wonCount` was missing you would have to access the player stats to sort the leaderboard.
+    * `addedAt` is a timestamp that indicates when the player's `wonCount` was last updated and determines the ranking when there is a tie in `wonCount` - the more recent, the closer to the `0` index in the array.
+
+    </HighlightBox>
+
+3. You make the `Leaderboard` message use it as an array. Add that each element in the map is not nullable. This will compile each `WinningPlayer` to a Go object instead of a pointer:
```

Comment:
@xavierlepretre 

> You make the `Leaderboard` message use **it** as an array.

"It" refers to "the Protobuf file"?

Reply:
No. Use `message Winner`.